src/box/lua/console.c

```diff
@@ -284,12 +309,35 @@ lbox_console_readline(struct lua_State *L)
 	while (top == lua_gettop(L)) {
 		while (coio_wait(STDIN_FILENO, COIO_READ,
 				 TIMEOUT_INFINITY) == 0) {
+			if (sigint_called) {
+				const char *line_end = "^C\n";
+				ssize_t rc = write(STDOUT_FILENO, line_end,
+						   strlen(line_end));
+				(void)rc;
+				/*
+				 * Discard current input and disable search
+				 * mode.
+				 */
+				RL_UNSETSTATE(RL_STATE_ISEARCH |
+					      RL_STATE_NSEARCH |
+					      RL_STATE_SEARCH);
+				rl_on_new_line();
+				rl_replace_line("", 0);
+				lua_pushstring(L, "");
+
+				readline_L = NULL;
+				sigint_called = false;
+				set_sigint_cb(old_cb);
+				return 1;
+			}
 			/*
 			 * Make sure the user of interactive
 			 * console has not hanged us, otherwise
 			 * we might spin here forever eating
 			 * the whole cpu time.
 			 */
+			if (fiber_is_cancelled())
+				set_sigint_cb(old_cb);
 			luaL_testcancel(L);
```

Comment:
Here (and at `fiber_cancel()` below) I have a doubt. Should not we set the `sigint_called` variable to `false`? I don't find a way to observe anything bad (can't start a console after cancelling its fiber), but this assumption looks a bit fragile.

@vr009, @locker, what do you think?

Reply:
If we don't yield (and we don't except for `coio_wait`), the callback can't be invoked to set the flag so we don't need to clear anything AFAIU.

Reply:
We don't drop it to `false` at starting of the function. So if (in future) there will be a way to start a console after cancelling a fiber that serves it, `coio_wait()` may exit with `sigint_called == true` after regular input.

My attempt to reproduce it is the following:

```lua
local fiber = require('fiber')
local console = require('console')

local console_fiber = fiber.create(function()
    pcall(console.start)
end)
console_fiber:cancel()

fiber.create(console.start)
```

But I got:

```
$ tarantool tmp/console-start-stop.lua
tarantool> LuajitError: builtin/box/console.lua:748: console is already started
```

OTOH, we'll just print one more prompt.

Reply:
> We don't drop it to false at starting of the function.

If you mean the function `lbox_console_readline`, the variable `sigint_called` get set to false at the start of the function.

Reply:
Ouch. Either I missed it or looked at the old version of the code. Okay so.